### PR TITLE
Coding agents updates

### PR DESCRIPTION
**Pull request by ashesfall**

### CODE REVIEW

**Comment by ashesfall** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/almostrealism/common/pull/226#pullrequestreview-4259886990)

File: flowtree/src/main/java/io/flowtree/jobs/DeduplicationRule.java [hunk 1]
```diff
@@ -44,6 +71,26 @@ protected List<String> extractItems(ClaudeCodeJob job) {
...
 
+    /**
+     * Returns {@code false} once the pass cap has been reached, in addition to
+     * the unchanged-set resolved condition inherited from {@link SetComparisonRule}.
+     */
+    @Override
+    public boolean isViolated(ClaudeCodeJob job) {
+        if (passCount >= maxPasses) return false;
+        return super.isViolated(job);
+    }
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/src/main/java/io/flowtree/jobs/DeduplicationRule.java]: Fixed. `isViolated` no longer short-circuits when the pass cap is reached. Instead, `getMaxRetries()` returns `maxPasses` so the enforcement framework uses the attempt counter (via `attempts < rule.getMaxRetries()`) to bound the correction loop. `isViolated` now accurately reflects actual violation state at all times — the cap is the framework's responsibility, not the rule's. New tests `deduplicationRuleCapStopsAfterDefaultTwoPasses`, `deduplicationRuleCapFiveAllowsFivePasses`, and `deduplicationRuleCapOneAllowsSinglePass` all assert that `isViolated` still returns `true` after the cap is reached when duplicates remain.

File: flowtree/src/main/java/io/flowtree/jobs/DeduplicationRule.java [hunk 2]
```diff
@@ -36,6 +42,27 @@
...
+    /**
+     * Creates a rule with an explicit pass cap.
+     *
+     * @param maxPasses maximum deduplication sessions before self-reporting satisfied;
+     *                  must be positive
+     */
+    DeduplicationRule(int maxPasses) {
+        this.maxPasses = maxPasses;
+    }
+
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/src/main/java/io/flowtree/jobs/DeduplicationRule.java:68]: Fixed. The constructor now validates `maxPasses > 0` and throws `IllegalArgumentException("maxPasses must be positive, got: " + maxPasses)` for any non-positive value. Tests `deduplicationRuleConstructorRejectsZero` and `deduplicationRuleConstructorRejectsNegative` cover both boundary cases.

File: flowtree/src/main/java/io/flowtree/jobs/ClaudeCodeJob.java [hunk 3]
```diff
@@ -528,6 +534,11 @@ public void setDeduplicationMode(String deduplicationMode) {
         this.deduplicationMode = deduplicationMode;
     }
 
+    /** Returns the per-job deduplication pass cap; defaults to {@link #DEFAULT_MAX_DEDUP_PASSES}. */
+    public int getMaxDeduplicationPasses() { return maxDeduplicationPasses; }
+    /** Sets the per-job deduplication pass cap (see {@link #DEFAULT_MAX_DEDUP_PASSES}). */
+    public void setMaxDeduplicationPasses(int maxDeduplicationPasses) { this.maxDeduplicationPasses = maxDeduplicationPasses; }
+
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/src/main/java/io/flowtree/jobs/ClaudeCodeJob.java]: Fixed. `setMaxDeduplicationPasses` now throws `IllegalArgumentException("maxDeduplicationPasses must be positive, got: " + maxDeduplicationPasses)` for any value `<= 0`. Tests `setMaxDeduplicationPassesRejectsZero` and `setMaxDeduplicationPassesRejectsNegative` verify this for both boundary values.

File: flowtree/src/main/java/io/flowtree/jobs/ClaudeCodeJob.java [hunk 4]
```diff
@@ -1541,6 +1555,9 @@ public void set(String key, String value) {
             case "dedupMode":
                 this.deduplicationMode = value;
                 break;
+            case "maxDedupPasses":
+                this.maxDeduplicationPasses = Integer.parseInt(value);
+                break;
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/src/main/java/io/flowtree/jobs/ClaudeCodeJob.java:1564]: Fixed. The deserialization path for `maxDedupPasses` is now guarded by a try/catch for `NumberFormatException` (handles null, empty, and non-numeric values) and an additional `p > 0 ? p : DEFAULT_MAX_DEDUP_PASSES` check for non-positive numeric values. Any malformed or invalid wire value silently falls back to `DEFAULT_MAX_DEDUP_PASSES`. Tests `maxDeduplicationPassesDeserializationFallsBackOnEmptyString`, `maxDeduplicationPassesDeserializationFallsBackOnInvalidString`, and `maxDeduplicationPassesDeserializationFallsBackOnNonPositive` cover all three malformed-input cases.

File: flowtree/src/main/java/io/flowtree/jobs/ClaudeCodeJobFactory.java [hunk 5]
```diff
@@ -608,6 +614,25 @@ public void setDeduplicationMode(String deduplicationMode) {
...
+    }
+
+    /**
+     * Sets the maximum number of deduplication passes for jobs created by this factory.
+     *
+     * @param maxDeduplicationPasses cap on correction sessions per job; must be positive
+     */
+    public void setMaxDeduplicationPasses(int maxDeduplicationPasses) {
+        this.maxDeduplicationPasses = maxDeduplicationPasses;
+        set("maxDedupPasses", String.valueOf(maxDeduplicationPasses));
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/src/main/java/io/flowtree/jobs/ClaudeCodeJobFactory.java:666]: Fixed. `ClaudeCodeJobFactory.setMaxDeduplicationPasses` now validates the input and throws `IllegalArgumentException("maxDeduplicationPasses must be positive, got: " + maxDeduplicationPasses)` for any value `<= 0`. The encoded property is only updated after validation succeeds. Test `factorySetMaxDeduplicationPassesRejectsZero` covers the zero boundary case.

File: flowtree/src/main/java/io/flowtree/jobs/ClaudeCodeJobFactory.java [hunk 6]
```diff
@@ -1007,6 +1033,11 @@ private void setEnforcementFlag(String key, String value) {
             case "dedupMode":
                 this.deduplicationMode = value;
                 return;
+            case "maxDedupPasses":
+                this.maxDeduplicationPasses = (value == null)
+                        ? ClaudeCodeJob.DEFAULT_MAX_DEDUP_PASSES
+                        : Integer.parseInt(value);
+                return;
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/src/main/java/io/flowtree/jobs/ClaudeCodeJobFactory.java:1089]: Fixed. The `setEnforcementFlag` case for `maxDedupPasses` in `ClaudeCodeJobFactory` now explicitly handles null and empty strings (falling back to `DEFAULT_MAX_DEDUP_PASSES`), wraps `Integer.parseInt` in a try/catch for `NumberFormatException`, and guards against non-positive parsed values with `(parsed > 0) ? parsed : DEFAULT_MAX_DEDUP_PASSES`. Tests `factoryMaxDeduplicationPassesDeserializationFallsBackOnEmptyString` and `factoryMaxDeduplicationPassesDeserializationFallsBackOnNonPositive` verify this behavior. All 78 `ClaudeCodeJobEnforcementTest` tests pass.

File: tools/mcp/secrets/server.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+                        ``workstream_id`` query parameter on every
+                        controller call and must match the workstream
+                        embedded in ``AR_MANAGER_TOKEN``.
+    AR_MANAGER_TOKEN    HMAC ``armt_tmp_*`` bearer token. The same
+                        token already issued for the job's ar-manager
+                        HTTP session is reused here.
+
+The server refuses to start when any of these is missing — running
+without authentication context would either expose all secrets or fail
+silently on every call.
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply tools/mcp/secrets/server.py:28]: Fixed. The module docstring at the top of `server.py` now accurately describes the actual behaviour: when a required environment variable is missing the server records a startup error and continues running, but every tool call returns `{"ok": false, "error": ...}` so callers receive a consistent error response rather than an unhandled exception.

File: tools/mcp/secrets/server.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+            tmp_fh.flush()
+            os.fsync(tmp_fh.fileno())
+        os.chmod(tmp_path, file_mode)
+        os.replace(tmp_path, expanded)
+    except Exception:
+        try:
+            os.remove(tmp_path)
+        except OSError:
+            pass
+        raise
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply tools/mcp/secrets/server.py:286]: Fixed. The filesystem write/rename block in `secret_render_file` now catches `Exception` instead of re-raising it. On failure it cleans up the temp file and returns `{"ok": False, "error": f"Failed to write secret file: {e}"}` so the MCP stdio server stays up and callers receive the standard error envelope.

File: tools/mcp/secrets/test_server.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    effective = dict(_DEFAULT_ENV)
+    if env is not None:
+        effective.update(env)
+    # Drop unset markers
+    effective = {k: v for k, v in effective.items() if v is not None}
+    with patch.dict(os.environ, effective, clear=False):
+        if "server" in sys.modules:
+            del sys.modules["server"]
+        import server as srv  # type: ignore
+        return srv
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply tools/mcp/secrets/test_server.py:65]: Fixed. `_import_server()` in `test_server.py` now uses `importlib.util.spec_from_file_location("ar_secrets_server", _SERVER_PATH)` to load the module under the unique name `ar_secrets_server` rather than the generic `server`. The old pattern deleted `sys.modules["server"]` before re-importing, which could clobber any other test file's `server` reference in the same pytest session. With the unique name, each load/reload is scoped to `ar_secrets_server` and cannot affect other modules.

File: docs/mcp/test_server.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    mcp_stdio.stdio_server = lambda: None
+    sys.modules.setdefault("mcp", mcp)
+    sys.modules.setdefault("mcp.server", mcp_server)
+    sys.modules.setdefault("mcp.types", mcp_types)
+    sys.modules.setdefault("mcp.server.stdio", mcp_stdio)
+
+
+_stub_mcp_imports()
+import server  # noqa: E402
+
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply docs/mcp/test_server.py:64]: Fixed. `docs/mcp/test_server.py` now loads the server module via `importlib.util.spec_from_file_location("ar_docs_server", ...)` under the unique name `ar_docs_server` instead of the generic `import server`. The `sys.path.insert` is removed; the module object is bound directly to the name `server` in module scope so all existing test assertions work unchanged. The `sys.modules` stub entries (`mcp`, `mcp.server`, etc.) are still installed via `setdefault` before the load, so they don't affect unrelated imports.

File: flowtree/src/main/java/io/flowtree/jobs/McpConfigBuilder.java [hunk 11]
```diff
@@ -186,9 +326,25 @@ public String buildMcpConfig() {
...
+        // by ManagedToolsDownloader and registered here as stdio entries.
+        Map<String, JsonNode> pushed = parsePushedConfig();
+        String homeDir = System.getProperty("user.home");
+        for (Map.Entry<String, JsonNode> entry : pushed.entrySet()) {
+            String name = entry.getKey();
+            ObjectNode serverNode = mcpServers.putObject(name);
+            serverNode.put("command", pythonCommand);
+            ArrayNode argsArray = serverNode.putArray("args");
+            argsArray.add(homeDir + "/.flowtree/tools/mcp/" + name + "/server.py");
+            JsonNode envNode = entry.getValue().get("env");
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/src/main/java/io/flowtree/jobs/McpConfigBuilder.java:340]: Fixed. `parsePushedConfig()` now validates every pushed server name against the pattern `^[a-z0-9][a-z0-9-]*$` (lowercase alphanumeric + hyphens, first char alphanumeric) before adding it to the result map. Names that don't match — including anything containing `/`, `..`, uppercase letters, spaces, or commas — are skipped with a `warn()` log. Since all callers of `parsePushedConfig()` (both `buildMcpConfig()` and `buildAllowedTools()`) receive the already-filtered map, the path used in `~/.flowtree/tools/mcp/{name}/server.py` can never escape the intended directory. New tests `parsePushedConfigRejectsInvalidServerNames` and `isValidServerNameAcceptsAndRejectsCorrectly` cover the pattern. All 14 `McpConfigBuilderTest` tests pass.

File: flowtree/src/main/java/io/flowtree/jobs/McpConfigBuilder.java [hunk 12]
```diff
@@ -226,11 +408,27 @@ public String buildAllowedTools(String baseTools) {
...
+        // Pushed-tool tools: tool names are listed verbatim in the JSON
+        Map<String, JsonNode> pushed = parsePushedConfig();
+        for (Map.Entry<String, JsonNode> entry : pushed.entrySet()) {
+            JsonNode toolsNode = entry.getValue().get("tools");
+            if (toolsNode == null || !toolsNode.isArray()) continue;
+            int count = 0;
+            for (JsonNode tn : toolsNode) {
+                if (tn.isTextual()) {
+                    sb.append(",mcp__").append(entry.getKey()).append("__").append(tn.asText());
+                    count++;
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/src/main/java/io/flowtree/jobs/McpConfigBuilder.java:476]: Fixed. `buildAllowedTools()` now validates each tool name against the pattern `^[a-z_][a-z0-9_]*$` (lowercase alphanumeric + underscores) before appending it to the `--allowedTools` string. Tool names containing commas, uppercase letters, spaces, or other characters that could alter the comma-separated string are skipped with a `warn()` log. Server names are already validated upstream in `parsePushedConfig()`, so both axes are protected. New tests `buildAllowedToolsRejectsInvalidToolNames` and `isValidToolNameAcceptsAndRejectsCorrectly` cover the pattern. All 14 `McpConfigBuilderTest` tests pass.

File: tools/mcp/secrets/server.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ashesfall) [reply tools/mcp/secrets/server.py:28]: Fixed. The module docstring no longer claims the server "refuses to start". It now accurately describes the actual behavior: when any required env var is missing, the startup error is captured in `_startup_error` and every tool call returns `{"ok": False, "error": ...}` describing the missing variable; the server process stays up so the agent can learn what went wrong via the MCP stdio handshake rather than experiencing a silent connection failure. Verified by `TestStartupConfig` (4/4 passing).

File: tools/mcp/secrets/server.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ashesfall) [reply tools/mcp/secrets/server.py:286]: Fixed. `secret_render_file` no longer re-raises on filesystem errors. The try block now spans `os.makedirs` + `tempfile.mkstemp` + the write/fsync/chmod/replace section, catches `(OSError, IOError)`, cleans up the temp file if one was created (`tmp_path` is initialized to `None` and reset to `None` on success), and returns `{"ok": False, "error": "Failed to write rendered file to <path>: <errno text>"}`. The error string is intentionally non-sensitive — it contains only the resolved output path and the OS-level exception text (errno + path), never any rendered payload value. Verified manually with a read-only target (`/proc/...`): server returns the error envelope cleanly instead of crashing. All 15 `tools/mcp/secrets/test_server.py` tests pass.

File: tools/mcp/secrets/test_server.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ashesfall) [reply tools/mcp/secrets/test_server.py:65]: Fixed. `_import_server` no longer touches the global `server` entry in `sys.modules`. Instead it loads the secrets server from its file path under the unique name `ar_secrets_server` via `importlib.util.spec_from_file_location("ar_secrets_server", server_path)` + `module_from_spec` + `loader.exec_module`. Subsequent reloads in the same test session drop only that specific key, never the generic `server` one. Confirmed via a direct probe: after `_import_server()`, `"server" in sys.modules` is `False` and `"ar_secrets_server" in sys.modules` is `True` — so a combined pytest run with manager/docs tests will no longer experience cross-module interference at the `server` sys.modules entry. All 15 secrets tests still pass.

File: flowtree/src/main/java/io/flowtree/slack/FlowTreeController.java [hunk 13]
```diff
@@ -1021,14 +1044,31 @@ private static Path resolveToolSource(String source, File configDir) {
      * the API endpoint reference and its resolved listening port.</p>
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/src/main/java/io/flowtree/slack/FlowTreeController.java:1047]: Fixed. The outdated "no longer actively used" sentence is gone. The Javadoc for `registerPushedTools()` now reads: "The resulting JSON is stored in `pushedToolsConfig` and forwarded to the API endpoint and Slack listener so that every agent job started by this controller can install and invoke the configured pushed tools (notably the built-in `ar-secrets` server). It is also surfaced via `getPushedToolsConfig()` for ops/debugging." This matches the actual control flow — the JSON is consumed by `ManagedToolsDownloader.ensurePushedTools` and `McpConfigBuilder.setPushedToolsConfig` in every job's launch path. `mvn clean install -DskipTests` is clean and 23/23 relevant tests (`ClaudeCodeJobPushedToolsTest`, `FlowTreeControllerMultiWorkspaceTest`, `PushedToolsSelfTestTest`) pass.
